pkg/ottl/functions.go

```diff
@@ -18,11 +18,38 @@ import (
 	"errors"
 	"fmt"
 	"reflect"
+	"strconv"
 	"strings"
 
+	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/pdata/pcommon"
 )
 
+// Arguments holds the arguments for an OTTL function, with arguments
+// specified as fields on a struct. Argument ordering is defined
+type Arguments interface{}
+
+// FunctionContext contains data provided by the Collector
+// component to the OTTL for use in functions.
+type FunctionContext struct {
+	Set component.TelemetrySettings
+}
+
+// Factory defines an OTTL function factory that will generate an OTTL
+// function to be called based on an invocation within a statement.
+type Factory[K any] interface {
+	// Name is the canonical name to be used by the user when invocating
+	// the function generated by this Factory.
+	Name() string
+
+	// CreateEmptyArguments initializes an Arguments struct specific to this
+	// Factory containing the arguments for the function.
+	CreateEmptyArguments() Arguments
```

Comment:
Default? To be consistent with components.

Reply:
I'm good with calling this `CreateDefaultArguments` if we have an express intent to support default function parameters using this function.

Thoughts @TylerHelmuth?

Reply:
I opened an issue to discuss optional parameters: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20879. I'll change the method name for now.

Reply:
Did you miss the name?

Reply:
Sorry, I took a look at something else and didn't push up the change. Everything should be renamed now.